File: app/src/main/java/com/simplemobiletools/filemanager/asynctasks/CopyTask.java
package com.simplemobiletools.filemanager.asynctasks;

import android.os.AsyncTask;
import android.support.v4.util.Pair;
import android.util.Log;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.ref.WeakReference;
import java.util.List;

public class CopyTask extends AsyncTask<Pair<List<File>, File>, Void, Boolean> {
    private static final String TAG = CopyTask.class.getSimpleName();

    private static WeakReference<CopyListener> mListener;
    private static File destinationDir;

    public CopyTask(CopyListener listener) {
        mListener = new WeakReference<>(listener);
    }

    @Override
    protected Boolean doInBackground(Pair<List<File>, File>... params) {
        final Pair<List<File>, File> pair = params[0];
        final List<File> files = pair.first;
        for (File file : files) {
            try {
                destinationDir = new File(pair.second, file.getName());
                copy(file, destinationDir);
                return true;
            } catch (Exception e) {
                Log.e(TAG, "copy " + e);
            }
        }
        return false;
    }

    private void copy(File source, File destination) throws Exception {
        if (source.isDirectory()) {
            if (!destination.exists() && !destination.mkdirs()) {
                throw new IOException("Could not create dir " + destination.getAbsolutePath());
            }

            final String[] children = source.list();
            for (String child : children) {
                copy(new File(source, child), new File(destination, child));
            }
        } else {
            final File directory = destination.getParentFile();
            if (directory != null && !directory.exists() && !directory.mkdirs()) {
                throw new IOException("Could not create dir " + directory.getAbsolutePath());
            }

            final InputStream in = new FileInputStream(source);
            final OutputStream out = new FileOutputStream(destination);

            final byte[] buf = new byte[1024];
            int len;
            while ((len = in.read(buf)) > 0) {
                out.write(buf, 0, len);
            }
            in.close();
            out.close();
        }
    }

    @Override
    protected void onPostExecute(Boolean success) {
        final CopyListener listener = mListener.get();
        if (listener == null)
            return;

        if (success) {
            listener.copySucceeded(destinationDir);
        } else {
            listener.copyFailed();
        }
    }

    public interface CopyListener {
        void copySucceeded(File destinationDir);

        void copyFailed();
    }
}

File: app/src/main/java/com/simplemobiletools/filemanager/asynctasks/CopyTask.kt
package com.simplemobiletools.filemanager.asynctasks

import android.content.Context
import android.os.AsyncTask
import android.support.v4.util.Pair
import android.util.Log
import com.simplemobiletools.filemanager.Utils
import com.simplemobiletools.filemanager.extensions.rescanItem
import com.simplemobiletools.filemanager.fragments.ItemsFragment
import java.io.*
import java.lang.ref.WeakReference
import java.util.*

class CopyTask(listener: CopyTask.CopyListener, val context: Context, val deleteAfterCopy: Boolean) : AsyncTask<Pair<List<File>, File>, Void, Boolean>() {
    private val TAG = CopyTask::class.java.simpleName
    private var mListener: WeakReference<CopyListener>? = null
    private var mMovedFiles: ArrayList<File>

    init {
        mListener = WeakReference(listener)
        mMovedFiles = arrayListOf()
    }

    override fun doInBackground(vararg params: Pair<List<File>, File>): Boolean? {
        val pair = params[0]
        val files = pair.first
        for (file in files) {
            try {
                val curFile = File(pair.second, file.name)
                if (curFile.exists())
                    continue

                copy(file, curFile)
            } catch (e: Exception) {
                Log.e(TAG, "copy $e")
                return false
            }
        }

        if (deleteAfterCopy) {
            for (file in mMovedFiles) {
                if (Utils.needsStupidWritePermissions(context, file.absolutePath)) {
                    Utils.getFileDocument(context, file.absolutePath).delete()
                } else {
                    file.delete()
                }
            }
        }
        return true
    }

    @Throws(Exception::class)
    private fun copy(source: File, destination: File) {
        if (source.isDirectory) {
            copyDirectory(source, destination)
        } else {
            copyFile(source, destination)
        }
    }

    private fun copyDirectory(source: File, destination: File) {
        if (!destination.exists()) {
            if (Utils.needsStupidWritePermissions(context, destination.absolutePath)) {
                val document = Utils.getFileDocument(context, destination.absolutePath)
                document.createDirectory(destination.name)
            } else if (!destination.mkdirs()) {
                throw IOException("Could not create dir ${destination.absolutePath}")
            }
        }

        val children = source.list()
        for (child in children) {
            val newFile = File(source, child)
            if (Utils.needsStupidWritePermissions(context, destination.absolutePath)) {
                if (newFile.isDirectory) {
                    copyDirectory(newFile, File(destination, child))
                } else {
                    var document = Utils.getFileDocument(context, destination.absolutePath)
                    document = document.createFile("", child)

                    val inputStream = FileInputStream(newFile)
                    val out = context.contentResolver.openOutputStream(document.uri)
                    copyStream(inputStream, out)
                    context.rescanItem(destination)
                    mMovedFiles.add(source)
                }
            } else {
                copy(newFile, File(destination, child))
            }
        }
    }

    private fun copyFile(source: File, destination: File) {
        val directory = destination.parentFile
        if (!directory.exists() && !directory.mkdirs()) {
            throw IOException("Could not create dir ${directory.absolutePath}")
        }

        val inputStream = FileInputStream(source)
        val out: OutputStream?
        if (Utils.needsStupidWritePermissions(context, destination.absolutePath)) {
            var document = Utils.getFileDocument(context, destination.absolutePath)
            document = document.createFile("", destination.name)
            out = context.contentResolver.openOutputStream(document.uri)
        } else {
            out = FileOutputStream(destination)
        }

        copyStream(inputStream, out)
        context.rescanItem(destination)
        mMovedFiles.add(source)
    }

    private fun copyStream(inputStream: InputStream, out: OutputStream?) {
        val buf = ByteArray(1024)
        var len: Int
        while (true) {
            len = inputStream.read(buf)
            if (len <= 0)
                break
            out?.write(buf, 0, len)
        }
    }

    override fun onPostExecute(success: Boolean) {
        val listener = mListener?.get() ?: return

        if (success) {
            listener.copySucceeded(if (deleteAfterCopy) ItemsFragment.ACTION_MOVE else ItemsFragment.ACTION_COPY)
        } else {
            listener.copyFailed()
        }
    }

    interface CopyListener {
        fun copySucceeded(action: Int)

        fun copyFailed()
    }
}
